### PR TITLE
fix: add test results aggregate utils

### PR DESCRIPTION
**Pull request by joseph-sentry**
depends on: https://github.com/codecov/shared/pull/282

Helper functions for test results aggregate queries

### CODE REVIEW

File: utils/test_results.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+from django.contrib.postgres.aggregates import ArrayAgg
+from django.db.models import Avg, Case, FloatField, Value, When
+from shared.django_apps.reports.models import Test, TestInstance
+
+thirty_days_ago = dt.datetime.now(dt.UTC) - dt.timedelta(days=30)
+
+
+def aggregate_test_results(
+    repoid: int, branch: str | None = None, history: dt.timedelta | None = None
+):
```
Comment (giovanni-guidini) [inline utils/test_results.py:12]: [nit] I'd love to see docstrings for this function (although reading through the code it's decently clear what it does. Good job with var naming

File: utils/test_results.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    average_duration_queryset = pass_failure_error_test_instances.values(
+        "test_id"
+    ).annotate(average_duration=Avg("duration_seconds"))
+
+    average_duration_dict = {
+        obj["test_id"]: obj["average_duration"] for obj in average_duration_queryset
+    }
+
+    tests = Test.objects.filter(repository_id=repoid).all()
+    result = dict()
```
Comment (giovanni-guidini) [inline utils/test_results.py:66]: [kinda nit] please create a [TypedDict](https://typing.readthedocs.io/en/latest/spec/typeddict.html#typeddict) or a class for this result type (and the typehint on the function return)

File: utils/tests/unit/test_test_results.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        create_test_instance_func(test_2, TestInstance.Outcome.SKIP, "c", duration=1)
+        create_test_instance_func(test_2, TestInstance.Outcome.ERROR, "d", duration=1)
+        create_test_instance_func(test_2, TestInstance.Outcome.PASS, "d", duration=1)
+        create_test_instance_func(test_2, TestInstance.Outcome.FAILURE, "d", duration=1)
+        create_test_instance_func(test_2, TestInstance.Outcome.FAILURE, "d", duration=1)
+        create_test_instance_func(test_2, TestInstance.Outcome.ERROR, "e", duration=1)
+
+        d = aggregate_test_results(repo_fixture.repoid)
+
+        assert d[test_1.id]["failure_rate"] == 0.5
```
Comment (giovanni-guidini) [inline utils/tests/unit/test_test_results.py:30]: can we have an "expected" dict and just compare both dicts? It'd be easier to understand what we actually expect :E

It also has the (small) added benefit to check that there's nothing extra that we don't expect in the response

**Review by giovanni-guidini — COMMENTED** [review]
In general I think it looks good, but left some comments

File: utils/test_results.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import datetime as dt
+
+from django.contrib.postgres.aggregates import ArrayAgg
+from django.db.models import Avg, Case, FloatField, Value, When
+from shared.django_apps.reports.models import Test, TestInstance
+
+thirty_days_ago = dt.datetime.now(dt.UTC) - dt.timedelta(days=30)
+
+
+def aggregate_test_results(
```
Comment (RulaKhaled) [inline utils/test_results.py:10]: How does this function work with pagination?

File: utils/test_results.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    average_duration_dict = {
+        obj["test_id"]: obj["average_duration"] for obj in average_duration_queryset
+    }
+
+    tests = Test.objects.filter(repository_id=repoid).all()
+    result = dict()
+    for test in tests:
+        result[test.id] = {
+            "failure_rate": failure_rate_dict.get(test.id),
```
[the remark this answers is not in the document]
Comment (joseph-sentry) [reply utils/test_results.py:69]: it currently just gets the aggregated fields, I can change it so it gets all the information relevant for display

File: utils/test_results.py (hunk 4 above)
Comment (joseph-sentry) [reply utils/test_results.py:10]: currently it does not, but i see that we'd want to paginate by test, like get the first 10 tests then the next 10 tests

**Pull request closed without merging**